test/jasmine/tests/transform_groupby_test.js

```diff
@@ -0,0 +1,523 @@
+            x: [1, -1, -2, 0, 1, 2, 3],
+            y: [1, 2, 3, 1, 2, 3, 1],
+
+            // everything is present:
+            transforms: [{
+                type: 'groupby',
+                groups: ['a', 'a', 'b', 'a', 'b', 'b', 'a'],
+                style: { a: {marker: {color: 'red'}}, b: {marker: {color: 'blue'}} }
+            }]
+        }];
+
+        var mockData0 = [{
+            mode: 'markers',
+            x: [1, -1, -2, 0, 1, 2, 3],
+            y: [1, 2, 3, 1, 2, 3, 1],
+
+            // groups, styles not present
+            transforms: [{
+                type: 'groupby'
+                // groups not present
+                // styles not present
+            }]
+        }];
+
+        // transform attribute with empty list
+        var mockData1 = [{
+            mode: 'markers',
+            x: [1, -1, -2, 0, 1, 2, 3],
+            y: [1, 2, 3, 1, 2, 3, 1],
+
+            // transforms is present but there are no items in it
+            transforms: [ /* list is empty */ ]
+        }];
+
+        // transform attribute with null value
+        var mockData2 = [{
+            mode: 'markers',
+            x: [1, -1, -2, 0, 1, 2, 3],
+            y: [1, 2, 3, 1, 2, 3, 1],
+            transforms: null
+        }];
+
+        // no transform is present at all
+        var mockData3 = [{
+            mode: 'markers',
+            x: [1, -1, -2, 0, 1, 2, 3],
+            y: [1, 2, 3, 1, 2, 3, 1]
+        }];
+
+        afterEach(destroyGraphDiv);
+
+        it('Plotly.plot should plot the transform traces', function(done) {
+            var data = Lib.extendDeep([], mockData);
+
+            var gd = createGraphDiv();
+
+            Plotly.plot(gd, data).then(function() {
+                expect(gd.data.length).toEqual(1);
+                expect(gd.data[0].x).toEqual([1, -1, -2, 0, 1, 2, 3]);
+                expect(gd.data[0].y).toEqual([1, 2, 3, 1, 2, 3, 1]);
+
+                expect(gd._fullData.length).toEqual(2); // two groups
+                expect(gd._fullData[0].x).toEqual([1, -1, 0, 3]);
+                expect(gd._fullData[0].y).toEqual([1, 2, 1, 1]);
+                expect(gd._fullData[1].x).toEqual([-2, 1, 2]);
+                expect(gd._fullData[1].y).toEqual([3, 2, 3]);
+
+                assertDims([4, 3]);
+
+                done();
+            });
+        });
+
+        // passes; maybe not for the good reasons (see fixme comments)
+        it('Plotly.plot should plot the transform traces', function(done) {
+            var data = Lib.extendDeep([], mockData0);
+
+            var gd = createGraphDiv();
+
+            Plotly.plot(gd, data).then(function() {
+                expect(gd.data.length).toEqual(1);
+                expect(gd.data[0].x).toEqual([1, -1, -2, 0, 1, 2, 3]);
+                expect(gd.data[0].y).toEqual([1, 2, 3, 1, 2, 3, 1]);
+
+                expect(gd._fullData.length).toEqual(0); // fixme: it passes with 0; shouldn't it be 1? (one implied group)
```

Comment:
I think it should also do the same as when `transforms.groups.length === 0`. Which leads to the question, what if `transforms.groups.length < x.length` and also, what we do now when arrays are of non-shared length. 

What the code _currently_ does is, it iterates through the elements of `groups` on `0 -> length - 1` i.e. the result will be empty.

While I can conditionalize it to do something differently (e.g. `if(transforms.groups === void(0) || transforms.groups.length === 0) return;` or some such) the user may pass e.g. one element in `groups` and gets another behavior.

@etpinard what's the desired outcome if 

1. the user doesn't specify a `transforms.groups`
2. the user _does_ specify a `transforms.groups` but it's `null`, `[]`, `undefined`
3. the user specifies a `transforms.groups` array but its length is shorter than the length of the (shortest?) data array

I'm almost leaning toward leaving the code as it is (i.e. no added logic), in favor of letting others work with `groupby` as early as possible, but it's probably because I'm not yet familiar with strong conventions in this realm so I'll gladly do what's deemed best.